nexus/src/app/background/tasks/blueprint_planner.rs

```diff
@@ -0,0 +1,209 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Background task for automatic update planning.
+
+use crate::app::background::BackgroundTask;
+use chrono::Utc;
+use futures::future::BoxFuture;
+use nexus_db_queries::context::OpContext;
+use nexus_db_queries::db::DataStore;
+use nexus_reconfigurator_planning::planner::Planner;
+use nexus_reconfigurator_preparation::PlanningInputFromDb;
+use nexus_types::deployment::{Blueprint, BlueprintTarget};
+use nexus_types::internal_api::background::BlueprintPlannerStatus;
+use serde_json::json;
+use std::sync::Arc;
+use tokio::sync::watch::{self, Receiver, Sender};
+
+/// Background task that runs the update planner.
+pub struct BlueprintPlanner {
+    datastore: Arc<DataStore>,
+    disabled: bool,
+    rx_blueprint: Receiver<Option<Arc<(BlueprintTarget, Blueprint)>>>,
+    tx_blueprint: Sender<Option<Arc<(BlueprintTarget, Blueprint)>>>,
+}
+
+impl BlueprintPlanner {
+    pub fn new(
+        datastore: Arc<DataStore>,
+        disabled: bool,
+        rx_blueprint: Receiver<Option<Arc<(BlueprintTarget, Blueprint)>>>,
+    ) -> Self {
+        let (tx_blueprint, _) = watch::channel(None);
+        Self { datastore, disabled, rx_blueprint, tx_blueprint }
+    }
+
+    pub fn watcher(
+        &self,
+    ) -> watch::Receiver<Option<Arc<(BlueprintTarget, Blueprint)>>> {
+        self.tx_blueprint.subscribe()
+    }
+
+    /// Run a planning iteration to generate a new blueprint.
+    /// If it is different from the current target blueprint,
+    /// save it and make it the current target.
+    pub async fn plan(&mut self, opctx: &OpContext) -> BlueprintPlannerStatus {
+        let mut status = BlueprintPlannerStatus::default();
+        if self.disabled {
+            debug!(&opctx.log, "blueprint planning disabled, doing nothing");
+            status.disabled = true;
+            return status;
+        }
+
+        // Get the current target blueprint to use as a parent.
```

Comment:
It's good that you're cloning here so that we don't block the channel.  I like to add a comment about that because I feel like it's a pretty non-obvious behavior about watch channels.

Reply:
Thanks, noted in 33d02e4fd.